core.go

```diff
@@ -500,13 +500,75 @@ func (m *Mat) ToImage() (image.Image, error) {
 			if channels == 4 {
 				c.A = uint8(data[y*step+x+3])
 			}
-			img.SetNRGBA(int(x/channels), y, c)
+			img.SetRGBA(int(x/channels), y, c)
 		}
 	}
 
 	return img, nil
 }
 
+//ImageToMatRGBA converts image.Image to gocv.Mat,
+//which represents RGBA image having 8bit for each component.
+//Type of Mat is gocv.MatTypeCV8UC4.
+func ImageToMatRGBA(img image.Image) (Mat, error) {
+	bounds := img.Bounds()
+	x := bounds.Dx()
+	y := bounds.Dy()
+	data := make([]byte, 0, x*y*4)
+	for j := bounds.Min.Y; j < bounds.Max.Y; j++ {
+		for i := bounds.Min.X; i < bounds.Max.X; i++ {
+			r, g, b, a := img.At(i, j).RGBA()
+			data = append(data, byte(b>>8), byte(g>>8), byte(r>>8), byte(a>>8))
+		}
+	}
+	return NewMatFromBytes(y, x, MatTypeCV8UC4, data)
+}
+
+//ImageToMatRGB converts image.Image to gocv.Mat,
+//which represents RGB image having 8bit for each component.
+//Type of Mat is gocv.MatTypeCV8UC3.
+func ImageToMatRGB(img image.Image) (Mat, error) {
+	bounds := img.Bounds()
+	x := bounds.Dx()
+	y := bounds.Dy()
+	data := make([]byte, 0, x*y*3)
+	for j := bounds.Min.Y; j < bounds.Max.Y; j++ {
+		for i := bounds.Min.X; i < bounds.Max.X; i++ {
+			r, g, b, _ := img.At(i, j).RGBA()
+			data = append(data, byte(b>>8), byte(g>>8), byte(r>>8))
+		}
+	}
+	return NewMatFromBytes(y, x, MatTypeCV8UC3, data)
+}
+
+//ImageToMatGray converts image.Image to gocv.Mat,
+//which represents grayscale image 8bit.
+//Type of Mat is gocv.MatTypeCV8UC1.
+func ImageToMatGray(img image.Image) (Mat, error) {
+	bounds := img.Bounds()
+	x := bounds.Dx()
+	y := bounds.Dy()
+	data := make([]byte, 0, x*y)
+	switch img.(type) {
+	case *image.Gray:
+		imgG := img.(*image.Gray)
+		for j := bounds.Min.Y; j < bounds.Max.Y; j++ {
+			for i := bounds.Min.X; i < bounds.Max.X; i++ {
+				data = append(data, imgG.GrayAt(i, j).Y)
+			}
+		}
+	default:
```

Comment:
The following code is intended to convert RGB to grayscale? Not sure if this is the best way to accomplish this, but if so it also needs test coverage.

Reply:
I had thoughts about it. There are different formula's. I think it's better to leave for user which way to choose.